server/src/user/auth.rs

```diff
@@ -0,0 +1,42 @@
+use crate::user::auth::AuthenticationError::{DatabaseError, UserNotFound};
+use crate::user::{User, UserData, UserError};
+use diesel::PgConnection;
+use pwhash::bcrypt;
+
+#[derive(Debug)]
+pub enum AuthenticationError {
+    IncorrectPassword,
+    UserNotFound,
+    BcryptError(pwhash::error::Error),
+    DatabaseError(UserError),
+}
+
+impl User {
+    pub fn generate_password(clear_password: &str) -> String {
+        bcrypt::hash(clear_password).unwrap()
+    }
+
+    fn check_password(self: &User, clear_password: &str) -> bool {
+        bcrypt::verify(clear_password, &*self.password)
+    }
+
+    pub fn authenticate(
+        conn: &PgConnection,
+        user_data: UserData,
+    ) -> Result<String, AuthenticationError> {
+        let user = User::find_by_username(&conn, &*user_data.username);
```

Comment:
Can we use the `?` here to propagate the error?

Reply:
This would require implementing a From for UserError -> AuthenticationError. I think the current way is better.